### PR TITLE
 feature:S3C-4408 store/retrieve secretkey on Redis w/ Encryption

### PR DESCRIPTION
**Pull request by saitharunthodupunuru**
Description:
* Store/retrieve credentials w/ Redis
* Implemented an external config file for cryptography
* Implemented AES256GCM encrypt/decrypt using nonce and associated data
* Implemented a CipherFactory to facilitate more cipher algorithms in future
* Encrypt/decrypt using cipherID and accessKey as associated data
* Serialize the encrypted data, cipher information used for the encryption (example: `nonce`), keyID into an object
* If a key is not available on Redis, its secret-key will be returned a `Not Available`
* If no keys present on Redis, a new key will automatically created during `listS3Credentials`
* Implement test cases

Design doc: https://github.com/scality/vmware-ose-scality/blob/development/1.0/Design.md

### CODE REVIEW

File: jacoco.gradle [hunk 1]
```diff
@@ -32,7 +32,9 @@ jacocoTestReport {
                     /*Ignore config classes*/
                     "**/com/scality/osis/configuration/*.*",
                     "**/com/scality/osis/*.*",
-                    "**/com/scality/osis/security/**/*.*"
+                    "**/com/scality/osis/security/basic/*.*",
+                    "**/com/scality/osis/utils/YamlPropertySourceFactory.*"
+
```
Comment (anurag4DSB) [inline jacoco.gradle:37]: left over extra line?

File: osis-core/src/main/java/com/scality/osis/utils/ScalityConstants.java [hunk 2]
```diff
@@ -30,6 +30,10 @@ private ScalityConstants() {
 
     public static final String FILTER_KEY_VALUE_SEPARATOR = "==";
 
+    public static final String REPO_KEY_SEPARATOR = "__";
+
+    public static final String REPO_VALUE_SEPARATOR = ";;";
```
Comment (anurag4DSB) [inline osis-core/src/main/java/com/scality/osis/utils/ScalityConstants.java:35]: Shouldn't this be `::`

**Review by anurag4DSB — APPROVED** [review]

File: osis-core/src/test/java/com/scality/osis/security/crypto/AES256GCMTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    @BeforeEach
+    public void setUp() {
+        aes256GCMUnderTest = new AES256GCM();
+    }
+
+    @Test
+    public void testEncryptDecrypt() throws Exception {
+        //create new random key
+        final String key = RandomStringUtils.random(32, 0, 0, true, true, null, new SecureRandom());
```
Comment (RachedBenMustapha) [inline osis-core/src/test/java/com/scality/osis/security/crypto/AES256GCMTest.java:30]: It's more of a personal preference matter, but I like to have hardcoded values in tests, to decrease the amount of moving parts.

**Review by ilkescality — APPROVED** [review]

**Review by RachedBenMustapha — COMMENTED** [review]
Didn't have enough time to go into details as much as I hoped.

Something I seem to be missing context on: since redis sentinel is not 100% as durable as eg Raft, there can be incidents where the last few redis ops are lost before the cluster re-converges. How would such a case be handled if some keys were missing from the redis cache (or the spring local hashmap)? When is the cache replenished?

File: osis-core/src/main/java/com/scality/osis/security/crypto/CryptoEnv.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@EnableConfigurationProperties(value = CryptoEnv.class)
+@ConfigurationProperties(prefix="osis.security")
+@PropertySources({
+        @PropertySource(value="classpath:crypto.yml", factory = YamlPropertySourceFactory.class),
```
[the remark this answers is not in the document]
Comment (saitharunthodupunuru) [reply osis-core/src/main/java/com/scality/osis/security/crypto/CryptoEnv.java:28]: Great point. Yes this was added for testing. I will remove it. Thanks

File: osis-core/src/main/java/com/scality/osis/utils/ScalityConstants.java (hunk 2 above)
Comment (saitharunthodupunuru) [reply osis-core/src/main/java/com/scality/osis/utils/ScalityConstants.java]: this was an unused variable. removed in 6a2f63423cce10658fa40363429704570b7528fb

File: osis-core/src/test/java/com/scality/osis/security/crypto/AES256GCMTest.java (hunk 3 above)
Comment (saitharunthodupunuru) [reply osis-core/src/test/java/com/scality/osis/security/crypto/AES256GCMTest.java]: Fixed in 6a2f63423cce10658fa40363429704570b7528fb. Great catch

File: src/main/resources/application.properties [hunk 5]
```diff
@@ -50,10 +50,10 @@ server.port=8443
 server.verbose=true
 
 # SSL Config
-server.ssl.key-store=file:/app/lib/osis.p12
-server.ssl.key-store-password=scality
+server.ssl.key-store=classpath:osis.p12
```
[the remark this answers is not in the document]
Comment (saitharunthodupunuru) [reply src/main/resources/application.properties]: This was an unnecessary change, fixed in 1388d02fde91cf6a133fdd548020a0bbffe6c1d1

File: osis-core/src/main/java/com/scality/osis/security/crypto/CryptoEnv.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (saitharunthodupunuru) [reply osis-core/src/main/java/com/scality/osis/security/crypto/CryptoEnv.java]: Fixed in 1388d02fde91cf6a133fdd548020a0bbffe6c1d1

File: osis-core/src/main/java/com/scality/osis/service/impl/ScalityOsisService.java [hunk 6]
```diff
@@ -12,6 +12,8 @@
 import com.amazonaws.services.securitytoken.model.Credentials;
 import com.amazonaws.util.StringUtils;
 import com.scality.osis.ScalityAppEnv;
+import com.scality.osis.redis.service.ScalityRedisRepository;
```
[the remark this answers is not in the document]
Comment (saitharunthodupunuru) [reply osis-core/src/main/java/com/scality/osis/service/impl/ScalityOsisService.java]: Fixed in 1388d02fde91cf6a133fdd548020a0bbffe6c1d1

File: jacoco.gradle (hunk 1 above)
Comment (saitharunthodupunuru) [reply jacoco.gradle]: removed in 6a2f634

**Comment by saitharunthodupunuru** [comment]
Added a new commit to handle latest design change. Commit: b6d9d10

Change:
1. If a key is not available on Redis, its `secret-key` will be returned a `Not Available`
2. If no keys present on Redis, a new key will automatically created during `listS3Credentials`

Design Commit: https://github.com/scality/vmware-ose-scality/commit/e1313ca46ea1990bf507879805325dc94f3b84c4

**Review by anurag4DSB — APPROVED** [review]

**Review by RachedBenMustapha — APPROVED** [review]

**Comment by saitharunthodupunuru** [comment]
Rebased with squash commits

**Pull request merged**